projects/client/RabbitMQ.Client/src/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -43,14 +43,16 @@ class WorkPool
             readonly ConcurrentQueue<Work> _workQueue;
             readonly CancellationTokenSource _tokenSource;
             readonly ModelBase _model;
-            readonly SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+            CancellationTokenRegistration _tokenRegistration;
+            volatile TaskCompletionSource<bool> _syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
             private Task _task;
 
             public WorkPool(ModelBase model)
             {
                 _model = model;
                 _workQueue = new ConcurrentQueue<Work>();
                 _tokenSource = new CancellationTokenSource();
+                _tokenRegistration = _tokenSource.Token.Register(() => _syncSource.TrySetCanceled());
```

Comment:
There is a closure allocation here but that is neglectable because the code is not on the hot path and only executed on stop